src/freenas/usr/local/bin/snmp-agent.py

```diff
@@ -437,19 +369,116 @@ def run(self):
+    row.setRowCell(9, agent.Counter64(io_overall[name]["read_bytes"]))
+    row.setRowCell(10, agent.Counter64(io_overall[name]["write_bytes"]))
+    row.setRowCell(11, agent.Counter64(io_1s[name]["read_ops"]))
+    row.setRowCell(12, agent.Counter64(io_1s[name]["write_ops"]))
+    row.setRowCell(13, agent.Counter64(io_1s[name]["read_bytes"]))
+    row.setRowCell(14, agent.Counter64(io_1s[name]["write_bytes"]))
+
+
+def fill_in_zvol_snmp_row_info(idx, zvolobj):
+    row = zvol_table.addRow([agent.Integer32(idx)])
+    row.setRowCell(1, agent.Integer32(idx))
+    row.setRowCell(2, agent.DisplayString(zvolobj.properties["name"].value))
+    row.setRowCell(3, agent.Counter64(int(zvolobj.properties["volsize"].rawvalue)))
+    row.setRowCell(4, agent.Counter64(int(zvolobj.properties["used"].rawvalue)))
+    row.setRowCell(5, agent.Counter64(int(zvolobj.properties["available"].rawvalue)))
+    row.setRowCell(6, agent.Counter64(int(zvolobj.properties["referenced"].rawvalue)))
+
+
+def report_zpool_and_zvol_info(prev_zpool_info, zfsobj):
+    zpool_table.clear()
+    zvol_table.clear()
+    for idx, zpool in enumerate(zfsobj.pools, start=1):
+        name = zpool.name
+
+        # zpool related information
+        io_overall, io_1s = gather_zpool_iostat_info(prev_zpool_info, name, zpool)
+        fill_in_zpool_snmp_row_info(idx, name, io_overall, io_1s, zpool)
+
+        # be sure and update our zpool io data so next time it's called
+        # we calculate the 1sec values properly
+        prev_zpool_info.update(io_overall)
+
+        # zvol related information
+        zvol_type = libzfs.DatasetType.VOLUME
+        for zvol in filter(lambda x: x.type == zvol_type, zpool.root_dataset.children_recursive):
+            # TODO: going through libzfs to get properties is expensive but there isn't a better
+            # way (currently) to get the relevant zvol information that we're looking for
+            fill_in_zvol_snmp_row_info(idx, zvol)
+
+
+def fill_in_dataset_snmp_row_info(idx, name, total, free, avail):
+    row = dataset_table.addRow([agent.Integer32(idx)])
+    row.setRowCell(1, agent.Integer32(idx))
+    row.setRowCell(2, agent.DisplayString(name))
+    row.setRowCell(3, agent.Counter64(total))
+    row.setRowCell(4, agent.Counter64(free))
+    row.setRowCell(5, agent.Counter64(avail))
+
+
+def report_dataset_info():
+    dataset_table.clear()
+    idx = 1
+    for devid, info in getmntinfo().items():
+        if info['fs_type'] != 'zfs':
+            continue
+
+        try:
+            st = os.statvfs(info['mountpoint'])
+        except Exception:
+            # don't crash, it's snmp so just continue
+            continue
+        else:
+            total_bytes = st.f_blocks * st.f_frsize
+            free_bytes = st.f_bfree * st.f_frsize
+            avail_bytes = st.f_bavail * st.f_frsize
+
+            idx += 1
+            fill_in_dataset_snmp_row_info(idx, info['mount_source'], total_bytes, free_bytes, avail_bytes)
```

Comment:
Does this mean the first row will have `idx=2`?

Reply:
Pushing a round of fixes which resolves this.